File: src/ex1.3.9.cpp
//Copyright (C) 2021  Brian Orwe
//
//This program is free software: you can redistribute it and/or modify
//it under the terms of the GNU General Public License as published by
//the Free Software Foundation, either version 3 of the License, or
//(at your option) any later version.
//
//This program is distributed in the hope that it will be useful,
//but WITHOUT ANY WARRANTY; without even the implied warranty of
//MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//GNU General Public License for more details.
//
//You should have received a copy of the GNU General Public License
//along with this program.  If not, see <http://www.gnu.org/licenses/>.


#include <iostream>
#include <string>
#include <cctype>
#include "stack.h"
#include "queue.h"


std::string normalize(std::string value_passed){
    std::string proper_sum;

    Stack<std::string> numbers;
    Stack<char> operators;

    for(char c:value_passed){
        if(std::isdigit(c)){
            numbers.push(std::string(1,c));
        }else if(c== '+'|| c== '-' || c=='*' || c=='/'){
            operators.push(c);
        }else if(c==')'){
            std::string right=numbers.pop();
            std::string left=numbers.pop();
            char ops= operators.pop();
            std::string result="("+left+ops+right+")";
            numbers.push(result);
        }
    }

    proper_sum=numbers.toString();

    return proper_sum;
}

int main(){
    std::string value_passed("1+2)*3-4)*5-6)))");
    std::cout<<"Turning "<<value_passed<<" to propert string:\n";
    std::string proper_sum=normalize(value_passed);
    std::cout<<"Done Result is: \n";
    std::cout<<proper_sum<<"\n";
    return 0;
}

File: src/main/java/com/borwe/algorithms/exercises/Ex1_3_9.java
/* 
 * Copyright (C) 2021 brian
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.borwe.algorithms.exercises;

import com.borwe.algorithms.algs.Queue;
import com.borwe.algorithms.algs.Stack;
import com.borwe.algorithms.algs.data.SimulateInput;

/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */

/**
 *
 * @author brian
 */
public class Ex1_3_9 {

	private static String
		 normalize(Queue<String> strings){
		Stack<String> operators=new Stack<>();
		Stack<String> operands=new Stack<>();

		for(String s:strings){
			switch(s){
				case "*":
				case "+":
				case "-":
				case "/": operators.push(s);break;
				case ")": {
					String operator=operators.pop();
					String val2=operands.pop();
					String val1=operands.pop();
					operands.push("( "+val1+" "+operator
						+" "+val2+" )");
				} break;
				default: operands.push(s);
			}
		}

		return operands.pop();
	}

	public static void main(String[] args) {
		//get input from user
		Queue<String> words=SimulateInput.getUserInputWords("Enter some words");

		System.err.println("INPUT:"+words.toString());
		System.err.println("OUPUT: "+normalize(words));

	}
	
}
